### PR TITLE
enable functions to simulate_design

### LINKED ISSUE #230: simulate a design based on a custom design function rather than design object
@macartan suggests allowing users to write their own design function that returns a run of the design (i.e. estimates and estimands)

This seems in line with our general philosophy of enter-and-exit-our-framework-at-will. Does not seem too hard to implement either. 

The function they write would have to return a list with two entries, estimates and estimands.

Others, thoughts?

### PR DESCRIPTION
**Pull request by graemeblair**
closes https://github.com/DeclareDesign/DeclareDesign/issues/230

you can run, simulate, and diagnose functions that return lists of estimates_df and estimands_df

### CODE REVIEW

File: R/diagnose_design.R [hunk 1]
```diff
@@ -104,10 +104,10 @@ diagnose_design <- function(...,
     } else {
       designs <- dots
     }
-    if (!all(vapply(designs, inherits, FALSE, "design"))) {
-      stop(
-        "Please only send design objects created using the + operator to diagnose_design."
-      )
+    if (!all(sapply(designs, function(x) {
```
Comment (nfultz) [inline R/diagnose_design.R]: pull this into a helper function.

File: R/simulate_design.R [hunk 2]
```diff
@@ -96,11 +99,8 @@ simulate_design <-
       SIMPLIFY = FALSE
     )
     
-    #if (length(designs) > 1) {
-      simulations_list <-
-        Map(cbind, design_label = names(simulations_list), simulations_list, stringsAsFactors = FALSE)
-    #}
-    
+    simulations_list <-
```
Comment (nfultz) [inline R/simulate_design.R]: join these lines

File: R/diagnose_design.R (hunk 1 above)
Comment (nfultz) [reply R/diagnose_design.R]: and also use it in simulate

File: R/design_helper_functions.R [hunk 3]
```diff
@@ -127,6 +127,14 @@ run_design_internal.default <- function(design, current_df=NULL, results=NULL, s
   
 }
 
+# for when the user sends a function that runs a design itself 
+#   to run_design (or simulate_design / diagnose_design above it)
+run_design_internal.function <- function(design) {
```
Comment (nfultz) [inline R/design_helper_functions.R]: should also update the generic on run_design_internal.default to be design, now that accept multiple types, and have default thow an error.

**Review by nfultz — DISMISSED** [review]
seems fine.

**Comment by graemeblair** [comment]
thx, addressed

**Pull request merged**